### PR TITLE
New table

### LINKED ISSUE #984: New Table 🎉
The tables have a new design. We'll take the opportunity to rewrite it from zero. 

# Design
image
Specs: [https://share.goabstract.com/902b44e5-a81e-43c7-b4e2-24eff8c88368](https://share.goabstract.com/902b44e5-a81e-43c7-b4e2-24eff8c88368)

# Legacy

We don't need to create a separate package. Both npm and yarn support [aliasing](https://stackoverflow.com/questions/26414587/how-to-install-multiple-versions-of-package-using-npm/56495651#56495651).

With that, we can start using the new table without the need to refactor all the views that use the old one.

# CSS Grid

Let's use the technique described in [this article](https://adamlynch.com/flexible-data-tables-with-css-grid/) and use the regular `<table>` component with `display: grid` to provide us more control over the table's behavior and style. It will [gracefully degrade](https://developer.mozilla.org/en-US/docs/Glossary/Graceful_degradation) to the regular table in old browsers (which we don't support but it's good to know it won't completely break).

# Dumb it down

Our current table component tries to do too many things at once, making it hard to extend and a mess to maintain.

My proposal is to follow the regular HTML Table structure. The components would be essentially styled-components of the original elements.
```jsx
import { Table, THead, TBody, TR, TH, TD } from "@crave/farmblocks-table";

export default () => (
  <Table>
    <THead>
      <TR>
        <TH>Fruit</TH>
        <TH>Price</TH>
        <TH></TH>
      </TR>
    </THead>
    <TBody>
      <TR>
        <TD>Apple</TD>
        <TD>$ 0.50</TD>
        <TD><button onClick={() => addToCart(0)}>Add to cart</button></TD>
      </TR>
      <TR>
        <TD>Banana</TD>
        <TD>$ 0.60</TD>
        <TD><button onClick={() => addToCart(1)}>Add to cart</button></TD>
      </TR>
      <TR>
        <TD>Cherry</TD>
        <TD>$ 9,999.99</TD>
        <TD><button onClick={() => addToCart(2)}>Add to cart</button></TD>
      </TR>
      <TR>
        <TD>Coconut</TD>
        <TD>$ 2.30</TD>
        <TD><button onClick={() => addToCart(3)}>Add to cart</button></TD>
      </TR>
    </TBody>
  </Table>
);
```

For ease of use, we'll also provide a helper to map the data from an array of objects:

```jsx
import {
  Table,
  THead,
  TBody,
  TR,
  TH,
  TD,
  useTableMap,
} from "@crave/farmblocks-table";

const fruits = [
  { id: 0, name: "Apple", price: "$ 0.50" },
  { id: 1, name: "Banana", price: "$ 0.60" },
  { id: 2, name: "Cherry", price: "$ 9,999.99" },
  { id: 3, name: "Coconut", price: "$ 2.30" },
];

export default () => {
  const columns = useMemo(() => ["name", "price", "actions"], []);
  const cellComponents = useMemo(() => {
    price: styled(TD)`font-weight: bold;`,
    actions: fruit => <TD><button onClick={() => addToCart(fruit.id)}>Add to cart</button></TD>
    _default: TD,
  }, []);
  const mapData = useTableMap({
    columns,
    keyProp: "id", // Optional but recommended. If not given, it will use the index.
    rowComponent: TR, // Optional. Could be used when we extend the component with styled(). Defaults to `TR`.
    cellComponents, // Optional. Defaults to `{_default: TD}`
    placeholder: "N/A", // Optional..
  });

  return (
    <Table>
      <THead>
        <TR>
          <TH>Fruit</TH>
          <TH>Price</TH>
          <TH></TH> {/* actions column, we don't want to label it */}
        </TR>
      </THead>
      <TBody>{fruits.map(mapData)}</TBody>
    </Table>
  );
};
```
We'll make the component as simple as possible and add features by creating helper functions like this one as we need.

### PR DESCRIPTION
**Pull request by vNakamura**
## Closes Issues

* Closes #984 

### Reviewer 1

* [ ] Functions properly
* [ ] Good readability, clarity and overall quality
* [ ] Has tests
* [ ] Has docs

### Reviewer 2

* [ ] Functions properly
* [ ] Good readability, clarity and overall quality
* [ ] Has tests
* [ ] Has docs

### CODE REVIEW

File: packages/table/src/hooks/useTableMap.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+            cellComponents[columnName] || cellComponents.$default || TD;
+
+          const content =
+            row[columnName] ||
+            placeholders[columnName] ||
+            placeholders.$default;
+
+          return (
+            <Cell
+              key={columnName}
```
Comment (alcferreira) [inline packages/table/src/hooks/useTableMap.js:29]: This `key` will repeat on every row... is it intended? 🤔

File: packages/table/src/components.styled.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+})`
+  border-bottom: 1px solid ${colors.GREY_08};
+  padding: 16px 24px;
+  font-size: 16px;
+  line-height: 24px;
+  color: ${colors.OYSTER};
+`;
+TD.defaultProps = {
+  truncate: true,
+};
```
Comment (alcferreira) [inline packages/table/src/components.styled.js:67]: Cool, as we are extending the `Text` component, we could use its default properties as well to set color, font size, and font weight. What do you think?

```jsx
export const TH = styled(Text).attrs(props => ({
  as: "th",
  fontWeight: fontWeights.SEMIBOLD,
  size: props.fontSize || fontSizes.MEDIUM,
}))`
  border: solid ${colors.GREY_08};
  border-width: 1px 0;
  padding: 8px 24px;
  line-height: 24px;
  background-color: ${colors.SUGAR};

  white-space: nowrap;
`

export const TD = styled(Text).attrs(props => ({
  as: "td",
  type: props.fontType || colors.OYSTER, // we could use fontColor instead
  size: props.fontSize || fontSizes.MEDIUM,
}))`
  border-bottom: 1px solid ${colors.GREY_08};
  padding: 16px 24px;
  line-height: 24px;
`;
TD.defaultProps = {
  truncate: true,
};
;
```

Comment (alcferreira) [reply packages/table/src/components.styled.js:67]: Usage:
```jsx
<TR>
  <TD fontColor='#2ecc71' fontSize={22}>Green Apple</TD>
  <TDRed fontWeight={300}>$ 0.50</TDRed>
</TR>
<TR>
  <TD>Banana</TD>
  <TDRed>$ 0.60</TDRed>
</TR>
```

image

File: packages/table/src/hooks/useTableMap.js (hunk 1 above)
Comment (vNakamura) [reply packages/table/src/hooks/useTableMap.js:29]: It won't be a problem because the rows have their own keys (which should be unique).

File: packages/table/src/components.styled.js (hunk 2 above)
Comment (vNakamura) [reply packages/table/src/components.styled.js:67]: ok, I think it can be even simpler with `defaultProps`.
